src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs

```diff
@@ -0,0 +1,353 @@
+    {
+        int currentPosition = 0;
+
+        this.resourceMetricStartIndex = currentPosition;
+
+        currentPosition += LengthAndTagSize;
+
+        // Serialize Resource
+        // TODO: Avoid serializing it multiple times.
+        SerializeResource(buffer, ref currentPosition, resource);
+
+        // TODO: Serialize schema_url field
+
+        // Serialize ScopeMetrics field
+        this.scopeMetricsStartIndex = currentPosition;
+        foreach (KeyValuePair<string, List<Metric>> metric in this.scopeMetrics)
+        {
+            if (metric.Value.Count > 0)
+            {
+                currentPosition = this.scopeMetricsStartIndex;
+
+                currentPosition += LengthAndTagSize;
+
+                // Serialize this meter/scope
+                this.SerializeScopeMetrics(buffer, ref currentPosition, metric.Key, metric.Value);
+            }
+        }
+    }
+
+    internal void SerializeScopeMetrics(byte[] buffer, ref int currentPosition, string scopeName, List<Metric> metrics)
+    {
+        // Serialize scope information
+        // TODO: Avoid serializing for each export.
+        SerializeInstrumentationScope(buffer, ref currentPosition, scopeName, metrics[0].MeterTags);
+
+        this.metricStartIndex = currentPosition;
+        foreach (Metric metric in metrics)
+        {
+            currentPosition = this.metricStartIndex;
+
+            currentPosition += LengthAndTagSize;
+
+            // Serialize metrics for the meter/scope
+            this.SerializeMetric(buffer, ref currentPosition, metric);
+        }
+
+        // TODO: Serialize schema_url field.
+    }
+
+    private void SerializeMetric(byte[] buffer, ref int currentPosition, Metric metric)
+    {
+        WriteInstrumentDetails(buffer, ref currentPosition, metric);
+
+        this.intrumentStartIndex = currentPosition;
+        switch (metric.MetricType)
+        {
+            case MetricType.LongSum:
+            case MetricType.LongSumNonMonotonic:
+                {
+                    currentPosition = this.intrumentStartIndex;
+
+                    currentPosition += LengthAndTagSize;
+
+                    // Write isMonotonic tag
+                    ProtobufSerializerHelper.WriteBoolWithTag(buffer, ref currentPosition, FieldNumberConstants.Sum_is_monotonic, metric.MetricType == MetricType.LongSum);
+
+                    // Write aggregationTemporality tag
+                    ProtobufSerializerHelper.WriteEnumWithTag(buffer, ref currentPosition, FieldNumberConstants.Sum_aggregation_temporality, metric.Temporality == AggregationTemporality.Cumulative ? 2 : 1);
+
+                    this.metricPointStartIndex = currentPosition;
+                    foreach (var metricPoint in metric.GetMetricPoints())
+                    {
+                        currentPosition = this.metricPointStartIndex;
+
+                        currentPosition += LengthAndTagSize;
+
+                        var sum = (ulong)metricPoint.GetSumLong();
```

Comment:
Why do we have to cast this to `ulong`?

Reply:
The proto defines this as `sfixed64` which should be a `long`: https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L679

Reply:
long and ulong both are encoded the same way. The bit representation does not change, the change would be on the decoding side on how it is interpreted (in terms of signed versus unsigned).